marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/mapping/mappingUILib.sjs

```diff
@@ -0,0 +1,126 @@
+const entityLib = require("/data-hub/5/impl/entity-lib.sjs");
+const consts = require("/data-hub/5/impl/consts.sjs");
+
+const modelArray = fn.collection(entityLib.getModelCollection()).toArray().map(entityModel =>{
+  return entityModel.toObject();
+});
+const responseArray = [];
+
+function getEntitiesForUI(targetEntityName){
+  let targetEntityModel = getEntityModel(targetEntityName);
+  const entityModelsReferringToTargetEntity = getEntityModelsWithPropertyThatRefersToTargetEntity(targetEntityModel, targetEntityName);
+
+  //Used to avoid cyclical dependencies
+  let entitiesNotToExpand = [targetEntityName];
+  getAllRelatedEntities(targetEntityName, targetEntityName, expandStructuredProperties(targetEntityModel, targetEntityName), entitiesNotToExpand,
+    {"mappingTitle": targetEntityName, "entityType": targetEntityName}, targetEntityName);
+  entityModelsReferringToTargetEntity.forEach(entityModel=>{
+    const entityName = entityModel.info.title;
+    getAllRelatedEntities(entityName, targetEntityName, expandStructuredProperties(entityModel, entityName), entitiesNotToExpand,
+      {"entityType": entityName}, entityName);
+  });
+  return responseArray;
+}
+
+function getAllRelatedEntities(entityName, targetEntityName, entityProperties, entitiesNotToExpand, entityResponseObject , propertyPath ) {
+  const entityModel = getEntityModel(entityName);
+  let expandedEntityModel = Object.assign({}, entityModel);
+  if(!entityResponseObject.entityModel){
+    expandedEntityModel.definitions = {};
+    expandedEntityModel.definitions[entityName] = {};
+    expandedEntityModel.definitions[entityName].properties = entityProperties;
+  }
+  if(!entityResponseObject.entityType){
+    entityResponseObject.entityType = entityName;
+  }
+
+  for (let entityPropertyName in entityProperties) {
```

Comment:
This for loop is fairly complex - multiple if conditionals, dozens of lines of code. See if the if blocks can be condensed into separate functions, where those functions have good names that document the condition they're handling.